quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java

```diff
@@ -124,15 +122,61 @@ public void drop(State state, Version version) throws MigrationException {
 		RefLog refLog = state.getRefLog();
 		Catalog catalog = state.getCatalog();
 
-		List<TableRef> tablesToDrop = refLog.getTableRefs().stream()
-				.filter(tableRef -> tableRef.getVersions().contains(version))
-				.filter(tableRef -> tableRef.getVersions().stream()
-						.filter(otherVersion -> otherVersion.getOperation().getType() == Type.DDL)
-						.count() == 1)
-				.collect(Collectors.toList());
+		Set<Version> activeVersions = refLog.getVersions();
+		checkArgument(activeVersions.contains(version), "Version to drop is not part of the active versions!");
+
+		// Check if version is at the end of the active versions
+		Version first = version;
+		Version last = version;
+		Version intermediateV = version.getChild();
+		while (intermediateV != null) {
+			if (activeVersions.contains(intermediateV)) {
+				last = intermediateV;
+			}
+			intermediateV = intermediateV.getChild();
+		}
+		// Or first
+		intermediateV = version.getParent();
+		while (intermediateV != null) {
+			if (activeVersions.contains(intermediateV)) {
+				first = intermediateV;
+			}
+			intermediateV = intermediateV.getParent();
+		}
+
+		checkArgument(!(first.equals(version) && last.equals(version)), "Cannot drop the only active version!");
+		checkArgument(first.equals(version) || last.equals(version), "The version to drop is not at the beginning or end of the active versions!");
```

Comment:
Is there a reason why this would be a requirement? If there are 3 schemas active at the same time (with or without intermediate version gaps), you should still be able to drop the middle version. This is actually used in some more complex migrations if I remember correctly.

Reply:
Please see my reply above.
I think that you use the intermediate version to get to the next versions, which is why you cannot just drop the middle part. If you have a migration scenario that includes dropping intermediate versions, could you send it to me?

Reply:
If you have a more complex migration with multiple steps for instance. This is also why the triggers form a fully connected graph and not a single line between versions:

Let's say you start with V1 being active, and are executing a migration that has multiple intermediate versions. This is roughly what the timeline would look like (excuse the terrible ASCII art).

```
Step 0: Baseline

V1


Step 1: Create intermediate version

V1 <----> V2


Step 2: Create intermediate version

V1 <----> V2 <----> V3
 ^                  ^
  \________________/


Step 3: Drop middle intermediate version

V1 <--------------> V3


Step 4: Create intermediate version

V1 <--------------> V3 <----> V4
 ^                            ^
  \__________________________/

Step 5: Drop middle intermediate version

V1 <------------------------> V4

Step 6: Create intermediate version

V1 <------------------------> V4 <----> V5
 ^                                      ^
  \____________________________________/

Step 7: Drop middle intermediate version

V1 <----------------------------------> V5
```

Things get even more complex when you try something like A/B testing different changes (not currently supported with the CLI itself, but the core package does support it). Imagine you're starting with a base version V1, and want to try out two different changes to V1 simultaneously. VA and VB both share a common parent, but they don't rely on each other.

```
   V1
  /  \
VA    VB
```

Because of this complexity QuantumDB constructs the triggers in such a way that a fully connected graph is formed between all (semi-)active versions. This would be pretty hard if you're relying on triggers triggering other triggers.